blacs_workers.py

```diff
@@ -180,20 +192,40 @@ def transition_to_buffered(self, device_name, h5file, initial_values, fresh):
         ext = self.device_properties['external_clock']
         freq = self.device_properties['clock_frequency']
         self.intf.send_command_ok(f"clk {ext:d} {freq:.0f}")
-            
+
+        # check if it is more efficient to fully refresh
+        if not fresh and self.smart_cache['do_table'] is not None:
+            total_inst = len(reps)
+            new_inst = 0
+            for i, (output, rep) in enumerate(zip(do_table, reps)):
+                if i >= len(self.smart_cache['reps']):
+                    new_inst += len(reps) - len(self.smart_cache['reps'])
+                    break
+                elif (self.smart_cache['do_table'][i] != output or
+                    self.smart_cache['reps'][i] != rep):
+                    new_inst += 1
+            if new_inst / total_inst > 0.1:
+                fresh = True
+
         # if fresh or not smart cache, program full table as a batch
         # this is faster than going line by line
         if fresh or self.smart_cache['do_table'] is None:
             print('programming from scratch')
             self.intf.send_command_ok('cls') # clear old program
-            self.intf.add_batch(do_table, reps)
+            self.intf.adm_batch(do_table, reps)
             self.smart_cache['do_table'] = do_table
             self.smart_cache['reps'] = reps
         else:
             print('incremental programming')
             # only program table lines that have changed
             for i, (output, rep) in enumerate(zip(do_table, reps)):
-                if (self.smart_cache['do_table'][i] != output or
+                if i >= len(self.smart_cache['reps']):

```

Comment:
Good catch here. I obviously did not consider the case when the total number of instructions have changed (which will likely be often).